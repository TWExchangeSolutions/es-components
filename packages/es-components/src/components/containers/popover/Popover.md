Popover utilizes a [render prop](https://reactjs.org/docs/render-props.html) to display the popover trigger, which is typically a button component.
This function will require `ref`, `toggleShow`, and `isOpen` parameters to function properly.

```
const styles = {
    display: 'inline-block',
    margin: '10px'
};

<div>
    <div style={styles}>
    <Popover
      name="topEx"
      title="Top"
      content="This is the popover's content. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch."
      placement="top"
      render={({ ref, toggleShow, isOpen }) => (
        <Button
          onClick={toggleShow}
          aria-expanded={isOpen}
          ref={ref}
          styleType="primary"
        >
          Popover on Top
        </Button>
      )}
    />
    </div>

    <div style={styles}>
    <Popover
        name="bottomEx"
        title="Bottom"
        content="This is the popover's content. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch."
        render={({ ref, toggleShow, isOpen }) => (
          <Button
            onClick={toggleShow}
            aria-expanded={isOpen}
            ref={ref}
            styleType="primary"
          >
            Popover on Bottom
          </Button>
        )}
      />
    </div>

    <div style={styles}>
    <Popover
        name="leftEx"
        title="Left"
        content="This is the popover's content. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch."
        placement="left"
        render={({ ref, toggleShow, isOpen }) => (
          <Button
            onClick={toggleShow}
            aria-expanded={isOpen}
            ref={ref}
            styleType="primary"
          >
            Popover on Left
          </Button>
        )}
      />
    </div>

    <div style={styles}>
    <Popover
        name="rightEx"
        title="Right"
        content="This is the popover's content. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch."
        placement="right"
        render={({ ref, toggleShow, isOpen }) => (
          <Button
            onClick={toggleShow}
            aria-expanded={isOpen}
            ref={ref}
            styleType="primary"
          >
            Popover on Right
          </Button>
        )}
      />
    </div>

</div>
```

```
<div>
Here's an example of a
<Popover
    name="popEx"
    title="Popover"
    content="This is the popover's content. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch."
    hasCloseButton
    render={({ ref, toggleShow, isOpen }) => (
      <PopoverLink
        onClick={toggleShow}
        aria-expanded={isOpen}
        ref={ref}
        styleType="primary"
      >
        Popover
      </PopoverLink>
    )}
  />
styled like a link.
</div>
```

Popovers can include Icons or other components. Keep accessibility in mind and provide aria labels where appropriate.

```
<div>
Click the icon for some help text.
<Popover
    name="iconEx"
    title="Icon Popover"
    content="Here's some help text!"
    suppressUnderline
    ariaLabel="Icon"
    render={({ ref, toggleShow, isOpen }) => (
      <PopoverLink
        onClick={toggleShow}
        aria-expanded={isOpen}
        ref={ref}
        suppressUnderline
        styleType="primary"
      >
        <span aria-hidden="true"><Icon name='question-circle' /></span>
        <span style={{ fontSize: '0', height: '1px', display: 'block', overflow: 'hidden' }}>Icon</span>
      </PopoverLink>
    )}
/>
</div>
```

```
const styles = {
    display: 'inline-block',
    margin: '0 10px'
};

<div>
    <div style={styles}>
    <Popover
        name="smEx"
        title="Small Arrow"
        content="This is the popover's content."
        placement="top"
        arrowSize="sm"
        render={({ ref, toggleShow, isOpen }) => (
          <OutlineButton
            onClick={toggleShow}
            aria-expanded={isOpen}
            ref={ref}
            styleType="primary"
          >
            Small Arrow
          </OutlineButton>
        )}
    />
    </div>

    <div style={styles}>
    <Popover
        name="dfltEx"
        title="Default Arrow"
        content="This is the popover's content."
        render={({ ref, toggleShow, isOpen }) => (
          <OutlineButton
            onClick={toggleShow}
            aria-expanded={isOpen}
            ref={ref}
            styleType="primary"
          >
            Default Arrow
          </OutlineButton>
        )}
    />
    </div>

    <div style={styles}>
    <Popover
        name="lgEx"
        title="Large Arrow"
        content="This is the popover's content."
        placement="top"
        arrowSize="lg"
        render={({ ref, toggleShow, isOpen }) => (
          <OutlineButton
            onClick={toggleShow}
            aria-expanded={isOpen}
            ref={ref}
            styleType="primary"
          >
            Large Arrow
          </OutlineButton>
        )}
    />
    </div>

    <div style={styles}>
    <Popover
        name="noneEx"
        title="No Arrow"
        content="This is the popover's content."
        placement="bottom"
        arrowSize="none"
        render={({ ref, toggleShow, isOpen }) => (
          <OutlineButton
            onClick={toggleShow}
            aria-expanded={isOpen}
            ref={ref}
            styleType="primary"
          >
            No Arrow
          </OutlineButton>
        )}
    />
    </div>
</div>
```
