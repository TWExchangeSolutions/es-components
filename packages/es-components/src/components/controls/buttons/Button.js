import React from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider, withTheme } from 'styled-components';

import defaultTheme from '../../theme/defaultTheme';

function getBlockPropertyValues(isBlock) {
  if (isBlock) {
    return {
      display: 'block',
      width: '100%'
    };
  }
  return {
    display: 'inline-block',
    width: 'initial'
  };
}

const ButtonBase = styled.button`
  border: none;
  border-radius: ${props => props.buttonSize.borderRadius};
  box-sizing: border-box;
  cursor: pointer;
  display: ${props => getBlockPropertyValues(props.block).display};
  font-family: inherit;
  font-size: ${props => props.buttonSize.fontSize};
  line-height: 1.4;
  padding: ${props => props.buttonSize.padding};
  text-align: center;
  vertical-align: middle;
  white-space: nowrap;
  width: ${props => getBlockPropertyValues(props.block).width};

  &[disabled] {
    cursor: not-allowed;
    opacity: 0.65;

    > * {
      pointer-events: none;
    }
  }
`;

const DefaultButton = styled(ButtonBase)`
  background-color: ${props => props.variant.bgColor};
  border-color: transparent;
  box-shadow: 0 4px 0 0 ${props => props.variant.boxShadowColor};
  color: ${props => props.variant.textColor};
  margin-bottom: 4px;
  margin-top: 0;

  &[disabled]:hover {
    background-color: ${props => props.variant.bgColor};
  }

  &:hover {
    background-color: ${props => props.variant.hoverBgColor};
  }

  &:active {
    margin-bottom: 0;
    margin-top: 4px;
    box-shadow: 0 0 0 0 transparent;
  }
`;

const OutlineButton = styled(ButtonBase)`
  background-color: ${props => props.variant.bgColor};
  border: 2px solid ${props => props.variant.borderColor};
  border-radius: 5px;
  color: ${props => props.variant.textColor};
  font-weight: bold;

  &:hover {
    background-color: ${props => props.variant.hoverBgColor};
    color: ${props => props.variant.hoverTextColor};
  }

  &:active {
    background-color: ${props => props.variant.activeBgColor};
    color: ${props => props.variant.activeTextColor};
  }
`;

const LinkButton = styled(ButtonBase)`
  background-color: transparent;
  box-shadow: none;
  color: ${props => props.variant.bgColor};
  font-size: inherit;
  padding: 0;
  text-decoration: underline;

  &:hover,
  &:focus,
  &:active {
    border-color: transparent;
    color: ${props => props.variant.hoverBgColor};
    text-decoration: none;
  }

  &:active {
    margin: 0 0 4px 0;
    box-shadow: none;
  }

  &[disabled]:hover {
    color: ${props => props.variant.bgColor};
    text-decoration: underline;
  }
`;

export function Button({
  handleOnClick,
  children,
  buttonClasses,
  styleType = 'default',
  isLinkButton = false,
  size = 'default',
  block = false,
  isOutline = false,
  theme,
  ...buttonProps
}) {
  const buttonSize = theme.buttonSizes[size];
  const sharedProps = {
    block,
    buttonSize,
    onClick: handleOnClick,
    className: buttonClasses,
    ...buttonProps
  };

  const defaultNormal = {
    bgColor: theme.colors.defaultColor,
    textColor: theme.colors.defaultBtnText,
    hoverBgColor: theme.colors.defaultHover,
    hoverTextColor: theme.colors.defaultBtnText,
    activeBgColor: theme.colors.defaultHover,
    activeTextColor: theme.colors.defaultBtnText,
    boxShadowColor: theme.colors.defaultHover
  };
  const defaultOutline = {
    bgColor: theme.colors.white,
    textColor: theme.colors.defaultColor,
    hoverBgColor: theme.colors.defaultColor,
    hoverTextColor: theme.colors.white,
    activeBgColor: theme.colors.defaultHover,
    activeTextColor: theme.colors.white,
    borderColor: theme.colors.defaultColor
  };

  let variant = theme.buttonStyles.buttonsNormal[styleType] || defaultNormal;
  let button = (
    <DefaultButton variant={variant} {...sharedProps}>
      {children}
    </DefaultButton>
  );

  if (isOutline) {
    variant = theme.buttonStyles.buttonsOutline[styleType] || defaultOutline;
    button = (
      <OutlineButton variant={variant} {...sharedProps}>
        {children}
      </OutlineButton>
    );
  } else if (isLinkButton) {
    variant = theme.buttonStyles.buttonsNormal[styleType] || defaultNormal;
    button = (
      <LinkButton variant={variant} {...sharedProps}>
        {children}
      </LinkButton>
    );
  }

  return <ThemeProvider theme={theme}>{button}</ThemeProvider>;
}

const buttonSizes = ['lg', 'default', 'sm', 'xs'];

Button.propTypes = {
  /** Function to execute when button is clicked */
  handleOnClick: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  /** Any additional classes to apply to the button */
  buttonClasses: PropTypes.string,
  /** Select the color style of the button, types come from theme */
  styleType: PropTypes.string,
  /** Determines if the button should be rendered as a text link */
  isLinkButton: PropTypes.bool,
  size: PropTypes.oneOf(buttonSizes),
  /** Make the button's width the size of it's parent container */
  block: PropTypes.bool,
  /** Render the outline button variant */
  isOutline: PropTypes.bool,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object,
  /** The name of the button to be sent with the form. */
  name: PropTypes.string
};

Button.defaultProps = {
  theme: defaultTheme
};

export default withTheme(Button);
