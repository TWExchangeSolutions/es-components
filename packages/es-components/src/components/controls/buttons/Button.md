Any additional props passed to the Button component will get passed to the HTML button element.

## Button Style Types

```
const buttonStyle = {
  display: 'inline-block',
  margin: '10px 15px 0 0'
};

<div>
  <div style={buttonStyle}><Button>Default</Button></div>

  <div style={buttonStyle}><Button styleType="darkDefault">Dark Default</Button></div>

  <div style={buttonStyle}><Button styleType="primary">Primary</Button></div>

  <div style={buttonStyle}><Button styleType="success">Success</Button></div>

  <div style={buttonStyle}><Button styleType="info">Information</Button></div>

  <div style={buttonStyle}><Button styleType="warning">Warning</Button></div>

  <div style={buttonStyle}><Button styleType="danger">Danger</Button></div>
</div>
```

## Button sizes

Including a size will set buttons to a specific size

```
const wrapperStyle = {
  marginBottom: '15px'
};

const buttonStyle = {
  marginRight: '15px'
};

<div>
  <div style={wrapperStyle}>
    <Button size="lg" style={buttonStyle} styleType="primary">Large</Button>
    <Button size="lg" style={buttonStyle}>Large</Button>
    <OutlineButton size="lg" style={buttonStyle} styleType="magenta">Large</OutlineButton>
  </div>

  <div style={wrapperStyle}>
    <Button style={buttonStyle} styleType="primary">Default</Button>
    <Button style={buttonStyle}>Default</Button>
    <OutlineButton style={buttonStyle} styleType="magenta">Default</OutlineButton>
  </div>

  <div style={wrapperStyle}>
    <Button size="sm" style={buttonStyle} styleType="primary">Small</Button>
    <Button size="sm" style={buttonStyle}>Small</Button>
    <OutlineButton size="sm" style={buttonStyle} styleType="magenta">Small</OutlineButton>
  </div>

  <div style={wrapperStyle}>
    <Button size="xs" style={buttonStyle} styleType="primary">Extra Small</Button>
    <Button size="xs" style={buttonStyle}>Extra Small</Button>
    <OutlineButton size="xs" style={buttonStyle} styleType="magenta">Extra Small</OutlineButton>
  </div>
</div>
```

Setting the `block` property will force the button to expand to the width of it's parent container.

```
const buttonContainerStyle = {
  width: '300px',
  padding: '10px',
  backgroundColor: '#ddd',
  border: '1px solid #444'
};

<div style={buttonContainerStyle}>
  <Button block styleType="primary">Block button</Button>
</div>
```

Any additional prop sent will be included on the button. For example, setting the `disabled` property will put the button into a disabled state, making it unclickable.

```
const buttonStyle = {
  display: 'inline-block',
  margin: '10px 15px 0 0'
};

<div>
  <div style={buttonStyle}><Button disabled>Default</Button></div>

  <div style={buttonStyle}><Button styleType="primary" disabled>Primary</Button></div>

  <div style={buttonStyle}><Button isOutline styleType="success" disabled>Success</Button></div>

  <div style={buttonStyle}><Button styleType="info" disabled>Information</Button></div>

  <div style={buttonStyle}><Button styleType="warning" disabled>Warning</Button></div>

  <div style={buttonStyle}><Button styleType="danger" disabled>Danger</Button></div>

  <div style={buttonStyle}><Button styleType="primary" isLinkButton disabled>Link</Button></div>
</div>
```
