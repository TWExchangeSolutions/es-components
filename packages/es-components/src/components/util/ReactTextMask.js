import React from 'react';
import ReactTextMask from 'react-text-mask';

// the component was using onInput which made testing more difficult.
// switched to onChange and put in a pull request so hopefully we can
// get rid of this soon.
ReactTextMask.prototype.render = function render() {
  const props = { ...this.props };

  delete props.mask;
  delete props.guide;
  delete props.pipe;
  delete props.placeholderChar;
  delete props.keepCharPositions;
  delete props.onChange;
  delete props.value;
  delete props.showMask;

  /* eslint-disable jsx-a11y/use-onblur-not-onchange */
  return (
    <input
      {...props}
      onChange={event => this.onChange(event)}
      defaultValue={this.props.value}
      ref={inputElement => (this.inputElement = inputElement)}
    />
    /* eslint-enable */
  );
};

export default ReactTextMask;
