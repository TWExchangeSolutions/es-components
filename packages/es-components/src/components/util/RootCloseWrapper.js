import React, { useRef } from 'react';
import { useRootClose } from 'react-overlays'

// eslint-disable-next-line react/prop-types
const RootCloseWrapper = ({children, onRootClose, disabled}) => {
    const ref = useRef();
    useRootClose(ref, onRootClose, {
        disabled
    });

    return (
      <div ref={ref} className="root-close-wrapper">
        {children}
      </div>
    );
}

export default RootCloseWrapper;