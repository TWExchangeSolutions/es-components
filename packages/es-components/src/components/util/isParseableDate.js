import parse from 'date-fns/parse';
import isValid from 'date-fns/is_valid';

export default function isParseableDate(date) {
  const parsedDate = parse(date);
  return isValid(parsedDate);
}
