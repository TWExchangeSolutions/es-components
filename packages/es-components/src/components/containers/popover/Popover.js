/* eslint-disable no-confusing-arrow */

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { findDOMNode } from 'react-dom';

import Icon from '../../base/icons/Icon';
import Button from '../../controls/buttons/Button';
import Popup from './Popup';

const Container = styled.div`
  display: inline-block;
`;

const TriggerButton = styled(Button)`
  border-bottom: ${props =>
    props.isLinkButton ? props.buttonBorderStyle : ''};
  border-radius: ${props =>
    props.isLinkButton && !props.suppressUnderline ? '0' : ''};
  margin-bottom: ${props => (props.isLinkButton ? '2px' : '')};
  text-decoration: none;

  &:hover {
    border-bottom: ${props =>
      props.isLinkButton && !props.suppressUnderline ? '1px solid' : ''};
  }
`;

const TriggerButtonLabel = styled.span`
  font-size: 0;
  height: 1px;
  overflow: hidden;
  display: block;
`;

const PopoverContainer = styled.div`
  min-width: 270px;
  max-width: 400px;
  background: ${props => props.theme.colors.white};
  border: 1px solid rgba(0, 0, 0, 0.2);
  box-shadow: 0 5px 10px rgba(0, 0, 0, 0.2);
`;

const PopoverHeader = styled.div`
  display: flex;
  justify-content: space-between;
  color: ${props => props.theme.colors.white};
  background-color: ${props =>
    props.hasTitle ? props.theme.colors.popoverHeader : 'none'};
  outline: none;
  line-height: ${props => props.theme.sizes.baseLineHeight};
`;

const TitleBar = styled.h3`
  font-size: 18px;
  padding: 8px 14px;
  margin: 0;
`;

const PopoverBody = styled.div`
  padding: ${props =>
    props.hasAltCloseWithNoTitle ? '0 14px 8px' : '8px 14px'};
  text-align: right;
  color: ${props => props.theme.colors.gray8};
  line-height: ${props => props.theme.sizes.baseLineHeight};
  font-size: 18px;
  font-weight: normal;
`;

const PopoverContent = styled.div`
  text-align: left;
  margin-bottom: ${props => (props.showCloseButton ? '8px' : '0')};
`;

const AlternateCloseButton = styled(Button)`
  color: ${props =>
    props.hasTitle ? props.theme.colors.white : props.theme.colors.grayDark};
  margin: 0;
  margin-left: auto;
  padding: 0 8px;
  background: transparent;
  border: none;
  box-shadow: none;

  &:hover {
    background: transparent;
  }
`;

const CloseHelpText = styled.span`
  height: 1px;
  width: 1px;
  color: transparent;
`;

class Popover extends React.Component {
  state = {
    isOpen: false
  };

  componentDidMount() {
    this.header = findDOMNode(this.headerRef);
    this.popoverContent = findDOMNode(this.contentRef);
    this.closeBtn = findDOMNode(this.closeBtnRef);
    this.triggerBtn = findDOMNode(this.triggerBtnRef);
    this.popper = findDOMNode(this.popperRef);
  }
  componentDidUpdate() {
    if (this.state.isOpen) {
      window.addEventListener('scroll', this.hidePopOnScroll);
    } else {
      window.removeEventListener('scroll', this.hidePopOnScroll);
    }
  }

  toggleShow = () => {
    setTimeout(() => {
      const focusableContent = this.popoverContent.querySelector('a, button');
      if (focusableContent) {
        focusableContent.focus();
      } else {
        this.header.focus();
      }
    }, 200);

    if (this.closeBtn) {
      this.triggerBtn.focus();
    }

    this.setState(({ isOpen }) => ({ isOpen: !isOpen }));
  };

  hidePopover = event => {
    if (this.state.isOpen) {
      if (event.type !== 'click') {
        this.triggerBtn.focus();
      }
      this.setState({ isOpen: false });
    }
  };

  hidePopOnScroll = () => {
    setInterval(() => {
      const bounds = this.popper.getBoundingClientRect();
      const inViewport =
        bounds.top >= 90 &&
        bounds.left >= 0 &&
        bounds.right <=
          (window.innerWidth || document.documentElement.clientWidth) &&
        bounds.bottom <=
          (window.innerHeight || document.documentElement.clientHeight);

      if (!inViewport && this.state.isOpen) {
        this.triggerBtn.focus();
        this.setState({ isOpen: false });
      }
    }, 100);
  };

  render() {
    const {
      name,
      title,
      content,
      children,
      placement,
      arrowSize,
      ariaLabel,
      buttonStyle,
      isOutline,
      isLinkButton,
      suppressUnderline,
      hasCloseButton,
      hasAltCloseButton,
      disableRootClose,
      disableFlipping
    } = this.props;

    const hasTitle = title !== undefined;
    const hasAltCloseWithNoTitle = !hasTitle && hasAltCloseButton;
    const showCloseButton = hasCloseButton && !hasAltCloseButton;
    const buttonBorderStyle = suppressUnderline ? 'none' : '1px dashed';

    const closeButton = (
      <Button
        handleOnClick={this.toggleShow}
        ref={btn => {
          this.closeBtnRef = btn;
        }}
      >
        Close
      </Button>
    );
    const altCloseButton = (
      <AlternateCloseButton
        aria-label="Close"
        hasTitle={hasTitle}
        handleOnClick={this.toggleShow}
        ref={btn => {
          this.closeBtnRef = btn;
        }}
      >
        <Icon name="remove" />
      </AlternateCloseButton>
    );

    const triggerButton = (
      <TriggerButton
        handleOnClick={this.toggleShow}
        styleType={buttonStyle}
        isOutline={isOutline}
        isLinkButton={isLinkButton}
        suppressUnderline={suppressUnderline}
        buttonBorderStyle={buttonBorderStyle}
        ref={btn => {
          this.triggerBtnRef = btn;
        }}
        aria-expanded={this.state.isOpen}
      >
        <span aria-hidden={!!ariaLabel}>{children}</span>
        <TriggerButtonLabel>{ariaLabel}</TriggerButtonLabel>
      </TriggerButton>
    );

    return (
      <Container className="es-popover">
        <Popup
          name={name}
          trigger={triggerButton}
          placement={placement}
          arrowSize={arrowSize}
          onHide={this.hidePopover}
          transitionIn={this.state.isOpen}
          hasTitle={hasTitle}
          disableRootClose={disableRootClose}
          disableFlipping={disableFlipping}
          popperRef={elem => {
            this.popperRef = elem;
          }}
        >
          <PopoverContainer
            className="es-popover__container"
            role="dialog"
            ref={elem => {
              this.contentRef = elem;
            }}
          >
            <PopoverHeader className="es-popover__header" hasTitle={hasTitle}>
              {hasTitle && <TitleBar>{title}</TitleBar>}
              {hasAltCloseButton && altCloseButton}
              <CloseHelpText
                tabIndex={-1}
                ref={elem => {
                  this.headerRef = elem;
                }}
                aria-label="Press escape to close the Popover"
              />
            </PopoverHeader>

            <PopoverBody
              className="es-popover__body"
              hasAltCloseWithNoTitle={hasAltCloseWithNoTitle}
            >
              <PopoverContent
                className="es-popover__content"
                showCloseButton={showCloseButton}
              >
                {content}
              </PopoverContent>
              {showCloseButton && closeButton}
            </PopoverBody>
          </PopoverContainer>
        </Popup>
      </Container>
    );
  }
}

Popover.propTypes = {
  /** The name of the popover. Used for differentiating popovers */
  name: PropTypes.string.isRequired,
  /** The content or element which activates the popover */
  children: PropTypes.node.isRequired,
  /** The text displayed in the popover title section */
  title: PropTypes.string.isRequired,
  /** The content displayed in the popover body */
  content: PropTypes.node.isRequired,
  /** The placement of the popover in relation to the link */
  placement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
  /** The size of the arrow on the popover box */
  arrowSize: PropTypes.oneOf(['sm', 'lg', 'none', 'default']),
  /** Prevents popover from closing when clicked outside of it */
  disableRootClose: PropTypes.bool,
  /** Display a close button in the bottom right of the popover body */
  hasCloseButton: PropTypes.bool,
  /** Display a close ('x') button in the popover title bar */
  hasAltCloseButton: PropTypes.bool,
  /** Hide underline from link. Useful for children like Icons */
  suppressUnderline: PropTypes.bool,
  /** The button style of the popover link */
  buttonStyle: PropTypes.string,
  /** Sets the link to use a text rather than a button style **/
  isLinkButton: PropTypes.bool,
  /** Sets the link to use the alternate button style **/
  isOutline: PropTypes.bool,
  /** Sets the aria-label attribute to allow for textless buttons */
  ariaLabel: PropTypes.string,
  /** Disables popovers ability to change position to stay in viewport */
  disableFlipping: PropTypes.bool
};

Popover.defaultProps = {
  placement: 'bottom',
  buttonStyle: 'primary'
};

export default Popover;
