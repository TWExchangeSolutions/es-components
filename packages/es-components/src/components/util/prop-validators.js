import isParseableDate from './isParseableDate';

export function isPropValidDate(props, propName, componentName) {
  const date = props[propName];
  if (date) {
    if (!isParseableDate(date)) {
      return new Error(`The pre-selected date is not valid. Please enter a parseable date.
        See http://en.wikipedia.org/wiki/ISO_8601 for valid date formats.
      `);
    }
    return null;
  }
  return null;
}
