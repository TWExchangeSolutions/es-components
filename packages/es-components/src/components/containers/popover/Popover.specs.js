/* eslint-env jest */

import React from 'react';
import { cleanup, waitForElement } from 'react-testing-library';

import Popover from './Popover';
import Button from '../../controls/buttons/Button';
import { renderWithTheme } from '../../util/test-utils';

beforeEach(cleanup);

function buildPopover(props) {
  const defaults = {
    name: 'popTest',
    title: 'Popover Title',
    content: 'This is the popover content.'
  };
  const mergedProps = Object.assign({}, defaults, props);
  return (
    <Popover
      {...mergedProps}
      renderTrigger={({ ref, toggleShow, isOpen }) => (
        <Button
          onClick={toggleShow}
          aria-expanded={isOpen}
          ref={ref}
          styleType="primary"
        >
          Popover Text
        </Button>
      )}
    />
  );
}

it('can be toggled by clicking the button', async () => {
  const { getByText, queryByText } = renderWithTheme(buildPopover());
  const trigger = getByText('Popover Text');

  trigger.click();
  const popoverContent = queryByText('This is the popover content.');
  expect(popoverContent).not.toBeVisible();

  trigger.click();
  expect(popoverContent).not.toBeVisible();
});

it('renders the title when provided', async () => {
  const { getByText, queryByText } = renderWithTheme(buildPopover());
  getByText('Popover Text').click();
  await waitForElement(() => getByText('Popover Title'));
  expect(queryByText('Popover Title')).not.toBeNull();
});

it('can be closed using the close button', async () => {
  const { getByText, queryByText } = renderWithTheme(
    buildPopover({ hasCloseButton: true })
  );
  getByText('Popover Text').click();

  const popoverContent = () => queryByText('This is the popover content.');
  popoverContent()
    .parentElement.querySelector('button')
    .click();

  await waitForElement(() => popoverContent() === null);
  expect(popoverContent()).toBeNull();
});

it('can be closed using the alternative close button', async () => {
  const { getByText, queryByText } = renderWithTheme(
    buildPopover({ hasAltCloseButton: true })
  );
  getByText('Popover Text').click();
  getByText('Popover Title')
    .parentElement.querySelector('button')
    .click();

  await waitForElement(
    () => queryByText('This is the popover content.') === null
  );
  expect(queryByText('This is the popover content.')).toBeNull();
});

it('sets focus on a focusable element within the content', () => {
  jest.useFakeTimers();
  const popoverContent = <a href="#test">Test link</a>;
  const { getByText } = renderWithTheme(
    buildPopover({ content: popoverContent })
  );
  getByText('Popover Text').click();
  jest.runOnlyPendingTimers();

  expect(getByText('Test link')).toHaveFocus();
});
