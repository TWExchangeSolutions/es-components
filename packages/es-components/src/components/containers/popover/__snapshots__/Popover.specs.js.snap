// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`popover component renders as expected 1`] = `
<div
  className="es-popover sc-bZQynM dSsQJY"
>
  <div>
    <div>
      <button
        aria-expanded={false}
        className="es-button es-button--default sc-gzVnrw iIXTKM sc-htpNat cJWpVN sc-bwzfXH jWFaYU"
        onClick={[Function]}
      >
        <span
          aria-hidden={false}
        >
          Popover Text
        </span>
        <span
          className="sc-htoDjs iqIjis"
        />
      </button>
    </div>
    <div
      className="sc-EHOje vChbq"
      name="popTest"
    >
      <div
        className="popTest-popper"
        data-placement={undefined}
        data-x-out-of-boundaries={undefined}
        style={
          Object {
            "opacity": 0,
            "pointerEvents": "none",
            "position": "absolute",
            "transition": "opacity 0.2s ease-in-out",
            "zIndex": 5,
          }
        }
      >
        <div
          className="es-popover__container sc-dnqmqq BnNnH"
          role="dialog"
        >
          <div
            className="es-popover__header sc-iwsKbI cirLJO"
          >
            <h3
              className="sc-gZMcBi cPBpHt"
            >
              Popover Title
            </h3>
            <span
              aria-label="Press escape to close the Popover"
              className="sc-fjdhpX dYNDTL"
              tabIndex={-1}
            />
          </div>
          <div
            className="es-popover__body sc-gqjmRU eiaKpk"
          >
            <div
              className="es-popover__content sc-VigVT cztEQg"
            >
              This is the popover content.
            </div>
          </div>
        </div>
        <span
          className="popTest-popper__arrow"
          style={Object {}}
        />
      </div>
    </div>
  </div>
</div>
`;
